File: ui/src/api/appstore.ts
import request from './request'

export interface AppTemplate {
  id?: number
  name: string
  display_name: string
  description: string
  icon: string
  code: string
  language: string
  category: string
  version: string
  author: string
  tags: string
  config_schema: string
  enabled: boolean
  created_at?: string
  updated_at?: string
}

export interface AppConfig {
  id?: number
  app_id: number
  key: string
  value: string
  type: string
  encrypted: boolean
  created_at?: string
  updated_at?: string
}

export interface InstallAppRequest {
  template_id: number
  name: string
  description: string
  user_id: number
  configs?: Record<string, string>
}

// 模板管理
export const getTemplates = (category?: string) => {
  return request.get<any, AppTemplate[]>('/appstore/templates', {
    params: { category }
  })
}

export const getTemplate = (id: number) => {
  return request.get<any, { data: AppTemplate }>(`/appstore/templates/${id}`)
}

export const createTemplate = (data: Partial<AppTemplate>) => {
  return request.post<any, { data: AppTemplate }>('/appstore/templates', data)
}

export const updateTemplate = (id: number, data: Partial<AppTemplate>) => {
  return request.put<any, { data: AppTemplate }>(`/appstore/templates/${id}`, data)
}

export const deleteTemplate = (id: number) => {
  return request.delete(`/appstore/templates/${id}`)
}

// 配置管理
export const getAppConfigs = (appId: number) => {
  return request.get<any, { data: AppConfig[] }>(`/appstore/configs/${appId}`)
}

export const createConfig = (data: Partial<AppConfig>) => {
  return request.post<any, { data: AppConfig }>('/appstore/configs', data)
}

export const updateConfig = (id: number, data: Partial<AppConfig>) => {
  return request.put<any, { data: AppConfig }>(`/appstore/configs/${id}`, data)
}

export const deleteConfig = (id: number) => {
  return request.delete(`/appstore/configs/${id}`)
}

// 从模板安装应用
export const installFromTemplate = (data: InstallAppRequest) => {
  return request.post('/appstore/install', data)
}

File: internal/app/models/appstore/appstore.go
package appstore

import "gorm.io/gorm"

type model struct {
	db *gorm.DB
}

func New(db *gorm.DB) Model {
	return &model{db: db}
}

// AppTemplate methods
func (m *model) CreateTemplate(template *AppTemplate) error {
	return m.db.Create(template).Error
}

func (m *model) GetTemplateByID(id uint) (*AppTemplate, error) {
	var template AppTemplate
	err := m.db.First(&template, id).Error
	if err != nil {
		return nil, err
	}
	return &template, nil
}

func (m *model) GetTemplateByName(name string) (*AppTemplate, error) {
	var template AppTemplate
	err := m.db.Where("name = ?", name).First(&template).Error
	if err != nil {
		return nil, err
	}
	return &template, nil
}

func (m *model) ListTemplates(category string) ([]*AppTemplate, error) {
	var templates []*AppTemplate
	query := m.db.Where("enabled = ?", true).Order("created_at DESC")
	if category != "" {
		query = query.Where("category = ?", category)
	}
	err := query.Find(&templates).Error
	return templates, err
}

func (m *model) UpdateTemplate(template *AppTemplate) error {
	return m.db.Save(template).Error
}

func (m *model) DeleteTemplate(id uint) error {
	return m.db.Delete(&AppTemplate{}, id).Error
}

// AppConfig methods
func (m *model) CreateConfig(config *AppConfig) error {
	return m.db.Create(config).Error
}

func (m *model) GetConfig(appID uint, key string) (*AppConfig, error) {
	var config AppConfig
	err := m.db.Where("app_id = ? AND key = ?", appID, key).First(&config).Error
	if err != nil {
		return nil, err
	}
	return &config, nil
}

func (m *model) GetConfigsByAppID(appID uint) ([]*AppConfig, error) {
	var configs []*AppConfig
	err := m.db.Where("app_id = ?", appID).Find(&configs).Error
	return configs, err
}

func (m *model) UpdateConfig(config *AppConfig) error {
	return m.db.Save(config).Error
}

func (m *model) DeleteConfig(id uint) error {
	return m.db.Delete(&AppConfig{}, id).Error
}

func (m *model) DeleteConfigsByAppID(appID uint) error {
	return m.db.Where("app_id = ?", appID).Delete(&AppConfig{}).Error
}
